exonum-java-binding-core/rust/src/storage/proof_map_index.rs

```diff
@@ -163,15 +173,75 @@ pub extern "system" fn Java_com_exonum_binding_storage_indices_ProofMapIndexProx
     let res = panic::catch_unwind(|| {
         let key = convert_to_key(&env, key)?;
         env.ensure_local_capacity(512)?;
-        let _proof = match *utils::cast_handle::<IndexType>(map_handle) {
+        let proof = match *utils::cast_handle::<IndexType>(map_handle) {
             IndexType::SnapshotIndex(ref map) => map.get_proof(key),
             IndexType::ForkIndex(ref map) => map.get_proof(key),
         };
-        unimplemented!("ECR-1350")
+        let unchecked_entries = proof.proof_unchecked();
+        let branches: Vec<_> = unchecked_entries.into_iter()
+            .map(|(path, hash)| create_java_map_proof_entry_branch(&env, path, hash))
+            .collect::<Result<Vec<_>>>()?;
+        let checked_proof = proof.check().unwrap();
+        let leaves = checked_proof.all_entries().into_iter().map(|(key, optional_value)| {
+            if optional_value.is_none() {
+                unimplemented!("Proofs for missing keys are not yet supported");
+            }
+            let path = ProofPath::new(key);
+            let value: Vec<u8> = optional_value.cloned().unwrap().into_bytes();
+            let leaf = create_java_map_proof_entry_leaf(&env, path, value);
+            leaf
+        }).collect::<Result<Vec<_>>>()?;
+        let array = create_java_array_map_proof_entry(&env, &leaves, &branches)?;
+        let unchecked_flat_map_proof = create_java_unchecked_flat_map_proof(&env, array)?;
+        Ok(unchecked_flat_map_proof.into_inner())
     });
     utils::unwrap_exc_or(&env, res, ptr::null_mut())
 }
 
+fn create_java_map_proof_entry_branch<'a>(env: &'a JNIEnv, proof_path: ProofPath, hash: Hash) -> Result<JObject<'a>> {
+    let proof_path: JObject = env.byte_array_from_slice(proof_path.as_bytes())?.into();
+    let hash: JObject = utils::convert_hash(&env, &hash)?.into();
+    env.new_object(
+        MAP_PROOF_ENTRY_BRANCH,
+        "([B[B)V",
+        &[proof_path.into(), hash.into()]
+    )
+}
+
+fn create_java_map_proof_entry_leaf<'a>(env: &'a JNIEnv, proof_path: ProofPath, value: Vec<u8>) -> Result<JObject<'a>> {
+    let proof_path: JObject = env.byte_array_from_slice(proof_path.as_bytes())?.into();
+    let value: JObject = env.byte_array_from_slice(&value)?.into();
+    env.new_object(
+        MAP_PROOF_ENTRY_LEAF,
+        "([B[B)V",
+        &[proof_path.into(), value.into()]
+    )
+}
+
+fn create_java_array_map_proof_entry<'a, 'b: 'a>(env: &'a JNIEnv, leaves: &'b [JObject], branches: &'b [JObject]) -> Result<JObject<'a>> {
+    let length = leaves.len() + branches.len();
+    let element_class = MAP_PROOF_ENTRY;
+    let array = env.new_object_array(length as i32, element_class, JObject::null())?;
+    let mut index = 0;
+    for branch in branches {
```

Comment:
@MakarovS , any ordering will do, is that correct?

Reply:
We need specific ordering, like stated in the algorithm here https://wiki.bf.local/display/EXN/Flat+proofs+for+ProofMapIndex
If we look at the proof from the example there, we need `H_1, V_1, H_2, H_3, H_4` ordering.

Reply:
thats not good tbh

Reply:
I think there is a misunderstanding: the algorithm checking the proof sorts the entries as its  first step, if I remember correctly. If that’s correct, we may use any order here. However, there might be some *preferred* order.

Reply:
Well, we can sort them either on Java or on Rust side. What do you choose? I think Rust would have better performance here.

Reply:
That’s a good question. Unless we say that a class representing a proof must get a sorted list of entries (i.e., it’s a precondition), it will have to always sort the input. I’m not sure we shall impose such a precondition - it might make unit testing a little more complex as we will have to always provide a sorted list of entries. Sorting of a sorted array will definitely be faster, but the compound operation (create and check) might not be.

Reply:
Hm, I expected proof to be passed to Java side already sorted and we check that it is.